app/services/control_rate_service.rb

```diff
@@ -10,60 +10,60 @@ def initialize(region, periods:)
     @region = region
     @facilities = region.facilities
     # Normalize between a single period and range of periods
-    @periods = if !periods.is_a?(Range)
+    @report_range = if !periods.is_a?(Range)
       # If calling code is asking for a single period,
       # we set the range to be the current period to the start of the next period.
       Range.new(periods, periods.succ)
     else
       periods
     end
-    @quarterly_report = @periods.begin.quarter?
-    @results = Reports::Result.new(@periods)
+    @quarterly_report = @report_range.begin.quarter?
+    @results = Reports::Result.new(region: @region, period_type: @report_range.begin.type)
     logger.info class: self.class, msg: "created", region: region.id, region_name: region.name,
-                periods: periods.inspect, facilities: facilities.map(&:id)
+                report_range: report_range.inspect, facilities: facilities.map(&:id)
   end
 
   delegate :logger, to: Rails
   attr_reader :facilities
-  attr_reader :periods
   attr_reader :region
+  attr_reader :report_range
   attr_reader :results
 
+  # We cache all the data for a region to improve performance and cache hits, but then return
+  # just the data the client requested
   def call
-    Rails.cache.fetch(cache_key, version: cache_version, expires_in: 7.days, force: force_cache?) do
-      results.registrations = registration_counts
-      results.cumulative_registrations = sum_cumulative_registrations
-      results.count_adjusted_registrations
-
-      periods.each do |(period, count)|
-        results.controlled_patients[period] = controlled_patients(period).count
-        results.uncontrolled_patients[period] = uncontrolled_patients(period).count
-      end
-
-      results.calculate_percentages(:controlled_patients)
-      results.calculate_percentages(:uncontrolled_patients)
-      results
-    end
+    all_cached_data.report_data_for(report_range)
   end
 
-  def sum_cumulative_registrations
-    earliest_registration_period = [periods.begin, registration_counts.keys.first].compact.min
-    (earliest_registration_period..periods.end).each_with_object(Hash.new(0)) { |period, running_totals|
-      previous_registrations = running_totals[period.previous]
-      current_registrations = registration_counts[period]
-      total = current_registrations + previous_registrations
-      running_totals[period] = total
+  private
+
+  def all_cached_data
+    Rails.cache.fetch(cache_key, version: cache_version, expires_in: 7.days, force: force_cache?) {
+      fetch_all_data
     }
   end
 
+  def fetch_all_data
+    results.registrations = registration_counts
```

Comment:
It does feel like the `Result` object is doing too much, at least purely from an API perspective (if not inherent behavior, like you [mentioned](https://github.com/simpledotorg/simple-server/pull/1428/files#r508985228))?

It can currently,

* Mutate data that it stores from the outside, by assigning stuff in the data hashmap (eg. `results.earliest_registration_period=`)
* Read data off the data hashmap by period (eg. `adjusted_registrations_for`)
* Read a different kind of data, that it computes (eg. `count_cumulative_registrations`)
* Read aggregates from the data it stores (eg. `results.calculate_percentages`)
* Rinse the data it currently holds (eg. `fill_in_nil_registrations`)

Reply:
I agree - my plan is to refactor towards a better, more focused Result object in subsequent PRs. This all got messy with the whole "cached result" vs "report result" differences that have come about.